### ISSUE TITLE
Thread priority inversion 

### ISSUE THREAD

**Issue by mHero21** [DeepARSDK/quickstart-ios-swift#28]
*Error Message:*
Thread running at QOS_CLASS_USER_INTERACTIVE waiting on a lower QoS thread running at QOS_CLASS_DEFAULT. Investigate ways to avoid priority inversions

* Reproducer:
1. Run Quickstart-ios-swift on Xcode 14(.3)
2. After initialisation you get this error

* Console log:
```
2023-05-24 14:44:14.458667+0200 quickstart-ios-swift[40795:2725925] Metal API Validation Enabled
Thread Performance Checker: Thread running at QOS_CLASS_USER_INTERACTIVE waiting on a lower QoS thread running at QOS_CLASS_DEFAULT. Investigate ways to avoid priority inversions
PID: 40795, TID: 2725925
Backtrace
=================================================================
3   DeepAR                              0x00000001052207ac _ZN2bx9Semaphore4waitEi + 64
4   DeepAR                              0x0000000105223a40 _ZN2bx6Thread4initEPFiPS0_PvES2_jPKc + 124
5   DeepAR                              0x000000010504f798 _ZN2ar10Controller10initializeENS_20DeepARInitParametersE + 448
6   DeepAR                              0x000000010505c588 _ZN2ar10DeepARImpl10initializeENS_20DeepARInitParametersE + 116
7   DeepAR                              0x000000010500232c -[DeepAR initializeWithWidth:height:window:] + 456
8   DeepAR                              0x0000000104ffda04 -[ARView initialize] + 564
9   DeepAR                              0x0000000105002638 -[DeepAR createARViewWithFrame:] + 92
10  quickstart-ios-swift                0x000000010422998c $s20quickstart_ios_swift14ViewControllerC20setupDeepARAndCamera33_1E468D19DB432620BB2613983BF78858LLyyF + 1340
11  quickstart-ios-swift                0x000000010422875c $s20quickstart_ios_swift14ViewControllerC11viewDidLoadyyF + 140
12  quickstart-ios-swift                0x0000000104228a00 $s20quickstart_ios_swift14ViewControllerC11viewDidLoadyyFTo + 36
13  UIKitCore                           0x000000018a94fdc4 179501B6-0FC2-344A-B969-B4E3961EBE10 + 3681732
14  UIKitCore                           0x000000018a5e7e40 179501B6-0FC2-344A-B969-B4E3961EBE10 + 110144
15  UIKitCore                           0x000000018a5e7b58 179501B6-0FC2-344A-B969-B4E3961EBE10 + 109400
16  UIKitCore                           0x000000018a7d41a4 179501B6-0FC2-344A-B969-B4E3961EBE10 + 2126244
17  UIKitCore                           0x000000018a7d4098 179501B6-0FC2-344A-B969-B4E3961EBE10 + 2125976
18  UIKitCore                           0x000000018a7d393c 179501B6-0FC2-344A-B969-B4E3961EBE10 + 2124092
19  UIKitCore                           0x000000018a9276dc 179501B6-0FC2-344A-B969-B4E3961EBE10 + 3516124
20  UIKitCore                           0x000000018a927034 179501B6-0FC2-344A-B969-B4E3961EBE10 + 3514420
21  UIKitCore                           0x000000018a925f88 179501B6-0FC2-344A-B969-B4E3961EBE10 + 3510152
22  UIKitCore                           0x000000018a925bd4 179501B6-0FC2-344A-B969-B4E3961EBE10 + 3509204
23  UIKitCore                           0x000000018a66b600 _UIScenePerformActionsWithLifecycleActionMask + 108
24  UIKitCore                           0x000000018a9ad918 179501B6-0FC2-344A-B969-B4E3961EBE10 + 4065560
25  UIKitCore                           0x000000018a85dfa4 179501B6-0FC2-344A-B969-B4E3961EBE10 + 2690980
26  UIKitCore                           0x000000018a85ddcc 179501B6-0FC2-344A-B969-B4E3961EBE10 + 2690508
27  UIKitCore                           0x000000018a85d97c 179501B6-0FC2-344A-B969-B4E3961EBE10 + 2689404
28  UIKitCore                           0x000000018a85d848 179501B6-0FC2-344A-B969-B4E3961EBE10 + 2689096
29  UIKitCore                           0x000000018af3ffa8 179501B6-0FC2-344A-B969-B4E3961EBE10 + 9908136
30  UIKitCore                           0x000000018afd8f98 _UISceneSettingsDiffActionPerformChangesWithTransitionContextAndCompletion + 224
31  UIKitCore                           0x000000018a709958 179501B6-0FC2-344A-B969-B4E3961EBE10 + 1296728
32  UIKitCore                           0x000000018ab7c7a8 179501B6-0FC2-344A-B969-B4E3961EBE10 + 5961640
33  UIKitCore                           0x000000018a7dc0b8 179501B6-0FC2-344A-B969-B4E3961EBE10 + 2158776
34  UIKitCore                           0x000000018a7dbf28 179501B6-0FC2-344A-B969-B4E3961EBE10 + 2158376
35  UIKitCore                           0x000000018a7db47c 179501B6-0FC2-344A-B969-B4E3961EBE10 + 2155644
36  UIKitCore                           0x000000018a7db208 179501B6-0FC2-344A-B969-B4E3961EBE10 + 2155016
37  FrontBoardServices                  0x000000019e035500 65E92D01-137F-3E16-98A6-A33706699695 + 13568
38  FrontBoardServices                  0x000000019e07451c 65E92D01-137F-3E16-98A6-A33706699695 + 271644
39  FrontBoardServices                  0x000000019e039294 65E92D01-137F-3E16-98A6-A33706699695 + 29332
40  FrontBoardServices                  0x000000019e074154 65E92D01-137F-3E16-98A6-A33706699695 + 270676
41  libdispatch.dylib                   0x000000010466e05c _dispatch_client_callout + 20
42  libdispatch.dylib                   0x0000000104671ad8 _dispatch_block_invoke_direct + 368
43  FrontBoardServices                  0x000000019e0433b0 65E92D01-137F-3E16-98A6-A33706699695 + 70576
44  FrontBoardServices                  0x000000019e042f4c 65E92D01-137F-3E16-98A6-A33706699695 + 69452
45  FrontBoardServices                  0x000000019e04572c 65E92D01-137F-3E16-98A6-A33706699695 + 79660
46  CoreFoundation                      0x00000001884e5f54 5CDC5D9A-E506-3740-B64E-BB30867B4F1B + 876372
47  CoreFoundation                      0x00000001884f232c 5CDC5D9A-E506-3740-B64E-BB30867B4F1B + 926508
48  CoreFoundation                      0x0000000188476210 5CDC5D9A-E506-3740-B64E-BB30867B4F1B + 418320
49  CoreFoundation                      0x000000018848bba8 5CDC5D9A-E506-3740-B64E-BB30867B4F1B + 506792
50  CoreFoundation                      0x0000000188490ed4 CFRunLoopRunSpecific + 612
51  GraphicsServices                    0x00000001c178e368 GSEventRunModal + 164
52  UIKitCore                           0x000000018a96f3d0 179501B6-0FC2-344A-B969-B4E3961EBE10 + 3810256
53  UIKitCore                           0x000000018a96f034 UIApplicationMain + 340
54  quickstart-ios-swift                0x0000000104231b44 main + 64
55  dyld                                0x00000001a6af8960 CB3FF411-4762-34D2-86A4-ECA13F9FB6C3 + 88416
Thread Performance Checker: Thread running at QOS_CLASS_USER_INTERACTIVE waiting on a lower QoS thread running at QOS_CLASS_DEFAULT. Investigate ways to avoid priority inversions
PID: 40795, TID: 2725925
Backtrace
=================================================================
3   DeepAR                              0x00000001052207ac _ZN2bx9Semaphore4waitEi + 64
4   DeepAR                              0x000000010537d854 _ZN4bgfx7Context11renderFrameEi + 292
5   DeepAR                              0x000000010537d61c _ZN4bgfx11renderFrameEi + 52
6   DeepAR                              0x000000010504fd10 _ZThn8_N2ar10Controller6renderEv + 24
7   QuartzCore                          0x0000000189ac6f9c 046338BF-09CF-3414-BC93-31296B97EB2D + 167836
8   QuartzCore                          0x0000000189ad88a4 046338BF-09CF-3414-BC93-31296B97EB2D + 239780
9   CoreFoundation                      0x000000018848b820 5CDC5D9A-E506-3740-B64E-BB30867B4F1B + 505888
10  CoreFoundation                      0x00000001884a8d00 5CDC5D9A-E506-3740-B64E-BB30867B4F1B + 625920
11  CoreFoundation                      0x00000001884aa908 5CDC5D9A-E506-3740-B64E-BB30867B4F1B + 633096
12  CoreFoundation                      0x000000018848c13c 5CDC5D9A-E506-3740-B64E-BB30867B4F1B + 508220
13  CoreFoundation                      0x0000000188490ed4 CFRunLoopRunSpecific + 612
14  GraphicsServices                    0x00000001c178e368 GSEventRunModal + 164
15  UIKitCore                           0x000000018a96f3d0 179501B6-0FC2-344A-B969-B4E3961EBE10 + 3810256
16  UIKitCore                           0x000000018a96f034 UIApplicationMain + 340
17  quickstart-ios-swift                0x0000000104231b44 main + 64
18  dyld                                0x00000001a6af8960 CB3FF411-4762-34D2-86A4-ECA13F9FB6C3 + 88416
```

**Comment by siliconprime-aunguyen** [comment]
I'm facing this issue too, any update?

**Comment by djanekovicdeepar** [comment]
Hey everyone,

we noticed and reproduced this issue in our tests but it didn't cause any obstructive effects. Are you noticing any unwanted artifacts due to this warning? I looked into both of these and why are they happening. I found that they are tied to deepar SDK initialization and I don't think they should cause any QoS degradation.

We most definitely have this on our dev roadmap and we will fix this eventually. From the initial triage, this didn't seem like a high priority issue.

Thanks,
Darko

**Comment by Kondamon** [comment]
This issue leads to crashes in production that are hard to find since the app just hangs and waits forever due to priority inversion. It's very easy to fix. You just need to change the priority of some tasks in your project. Never start a higher priority task within a low priority one.

**Comment by igor-aradski** [comment]
There were several fixes regarding this or similar issues. I'm closing the issue, but if you are experiencing the issue please leave a comment.
